### PR TITLE
Bug 2035534: Baremetal skip inspection

### PR DESCRIPTION
**Pull request by andfasano**
This patch adds a new test for the baremetal ipi platform, in order to verify that the inspection is skipped. Also, it adds the BaremetalTestHelper type (and also refactors a little bit the code) to start using the feature introduced in openshift-metal3/dev-scripts#1315.
The existing scale test has been also retouched to adopt the new add worker mechanism in case of baremetal ipi platform

### CODE REVIEW

**Comment by andfasano** [comment]
/hold

Requires https://github.com/openshift/origin/pull/26680

**Comment by andfasano** [comment]
cc @ardaguclu @bfournie @elfosardo

**Comment by andfasano** [comment]
/hold cancel

Dependency on #26680 has been removed

File: test/extended/baremetal/helper.go
Comment (ardaguclu) [inline test/extended/baremetal/helper.go:93]: Isn't there UnmarshallYaml function and that's why we are using `YAMLToJSON`?

File: test/extended/baremetal/helper.go
Comment (ardaguclu) [inline test/extended/baremetal/helper.go:155]: Isn't 2 minutes low for that?

File: test/extended/baremetal/helper.go
Comment (ardaguclu) [inline test/extended/baremetal/helper.go:187]: Clean up is already done in here https://github.com/openshift/origin/pull/26681/files#diff-7e730e2c7566572f6d8e0ef47a5b3ca546de7f0d1bd9f0468ff6bed4064a037cR140. Moreover, it waits deletion is completed. 

Is there any possibility that is executed?

File: test/extended/baremetal/helper.go
Comment (ardaguclu) [inline test/extended/baremetal/helper.go:149]: It looks like there is no point that calling `CreateExtraWorker` and not calling this function. Maybe, we can use it in `CreateExtraWorker` instead giving call responsibility to test owner?

**Review by ardaguclu — COMMENTED** [review]
Thanks @andfasano. These changes look great. I left some comments and I have one question;

We have to wait changes in `Machine Scale` test to merge this PR?, because it still assumes that machine is available.

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:31]: Seems that this should be called extraWorkers

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:133]: Maybe could use o.Eventually() here to be more consistent with gomega

File: test/extended/baremetal/helper.go
Comment (bfournie) [reply test/extended/baremetal/helper.go:155]: Same comment with using o.Eventually()

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:93]: Yes

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:149]: I didn't want to overengineer the API of the helper too much for the moment, and considered that the WaitForProvisioningState could be a more generic function useful for other tests (and other scenarios). I expect we're going to refine the interface as long as we're using it in more tests

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:187]: This block is in reality independent, and just required to support the migration, given that there could be other tests using the APPLY_EXTRA_WORKERS feature. Once we'll get rid of it, we'll remove it

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:155]: @bfournie I tried to use `Eventually()` but for this case it doesn't seem a better choice than `wait.PollIImmediate`: the comparison (on the provisioning state) must be moved outside the block, and also the timeouts must be specified at the end. In addition, the error check is present but must be positional (without the possibility to specify a message if required).
So the entire logical flow appeared a little bit strange, while in the current shape is anyhow pretty linear (also, `wait.poll...` seems to be a used more frequently in the repo rather than `Eventually`)

Comment (bfournie) [reply test/extended/baremetal/helper.go]: OK, I actually like the timeouts being specified at the end, but not being able to specify an error message seems like a limitation. Anyway, thanks for checking it out.

**Comment by andfasano** [comment]
/hold

Found a problem during the `Setup` cleanup logic, fixing it

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:81]: Nit - comment should be `getExtraWorkerData`

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:122]: Does anything need to be done to delete the secret that was created for the worker?

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:139]: Just curious, this will retry every second, so if there are any problems accessing the ApiServer on any of the retries the test will fail. It sounds like this will be fragile, but I guess this is how other tests work?

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:132]: Would be good to have a log message here, similar to waitForProvisioningState()

File: test/extended/baremetal/hosts.go
Comment (bfournie) [inline test/extended/baremetal/hosts.go:156]: Maybe change log msg - "Check that hardware field in status is emptry"

File: test/extended/baremetal/hosts.go
Comment (bfournie) [inline test/extended/baremetal/hosts.go:159]: So hardware is a pointer to a HardwareDetails struct, shouldn't there be a check for the value being nil instead of it being found?

File: test/extended/baremetal/commons.go
Comment (andfasano) [reply test/extended/baremetal/commons.go:96]: there's a name clash with the inner func but agree

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:122]: Secrets will be automatically garbage collected when the related host is deleted, so no need to explicitly manage them. Anyhow I found a case where the secret was not yet deleted, and the subsequent test started, and since this could cause a problem (the resource already exist) better to wait for their deletion as well

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:139]: In theory it's right, but  if we have issues in the apiserver in a serial test (so nothing else is running) then the situation is probably even worse :). Another strategy could be to filter other kinds of errors, it's probably something we could tweak in future in case of need (other tests adopted also the approach to watch for a particular event, but this one seems better to me)

File: test/extended/baremetal/hosts.go
Comment (andfasano) [reply test/extended/baremetal/hosts.go:159]: We don't have a structured type, so if not present it does not get unmarshalled at all (and thus the need to check the bool flag)

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:132]: Yep, good idea

**Review by elfosardo — APPROVED** [review]

**Comment by ardaguclu** [comment]
`e2e-metal-ipi-serial` failed due to the test in this PR, it looks like some fix needed.

**Comment by andfasano** [comment]
> `e2e-metal-ipi-serial` failed due to the test in this PR, it looks like some fix needed.

Yep, still working on it ([hold](https://github.com/openshift/origin/pull/26681#issuecomment-990988074) not yet removed)

**Comment by andfasano** [comment]
Requires https://github.com/metal3-io/baremetal-operator/pull/1054 (otherwise the deletion of extraworkers will intermittently fail)

**Comment by bfournie** [comment]
Seeing failure here which shouldn't be affected by changes:
```
Failing tests:

[sig-cluster-lifecycle][Feature:Machines][Serial] Managed cluster should grow and decrease when scaling different machineSets simultaneously [Suite:openshift/conformance/serial]
```

Also looks like a bugzilla needs to be added.

File: test/extended/baremetal/helper.go
Comment (andfasano) [reply test/extended/baremetal/helper.go:164]: Added `DeployExtraWorker`

File: test/extended/machines/scale.go
Comment (ardaguclu) [inline test/extended/machines/scale.go:208]: I think it would be better to do these operations only if platformType is equal to Baremetal. I know `NewBaremetalTestHelper` is intelligent enough to skip for other platform types but from this test's perspective, test just can not rely on it.

File: test/extended/machines/scale.go
Comment (ardaguclu) [inline test/extended/machines/scale.go:216]: I know helper in here is actually `NewBaremetalTestHelper` but reader can not easily understand this operation is also for other platforms?

File: test/extended/machines/scale.go
Comment (ardaguclu) [inline test/extended/machines/scale.go:197]: IMHO, instead wrapping up with BeforeEach and AfterEach just for Baremetal, I think it would be more convenient to use if baremetal and defer statements. Because in reality, this change has no effect on other platforms.

File: test/extended/machines/scale.go
Comment (ardaguclu) [inline test/extended/machines/scale.go:30]: As clearly seen you intentionally increased it. I'm asking just because of curiosity; what is the reason that?

**Review by ardaguclu — COMMENTED** [review]
Thanks @andfasano I left some comments

File: test/extended/machines/scale.go
Comment (andfasano) [reply test/extended/machines/scale.go:30]: In my local test I started to have timeout failure, as it seems that the provisioning time somehow slightly increased, so I also retouched it to see it if it works in the CI

File: test/extended/machines/scale.go
Comment (andfasano) [reply test/extended/machines/scale.go:197]: BeforeEach/AfterEach are the ways supported by [ginkgo](https://onsi.github.io/ginkgo/) for structuring the setup/teardown in a clear/readable mannder, and also ensuring the proper order of execution. I'd find complicated to follow other custom approaches for that (the helper teardown will not have any effect in case of non-baremetal platform)

File: test/extended/machines/scale.go
Comment (andfasano) [reply test/extended/machines/scale.go:208]: Considering that we are slightly enlarging the scopes of baremetal related operators (ie CBO) to other platforms, my intention was to have a more granular helper class. If we could have the same support on other platforms in future, this test could continue to work. Also, the setup/teardown code would be less cumbersome (across different tests), given that the helper will not required it (personally, I'd limit the platform type check only to decide if a whole test must be skipped or not).
I'm not very happy of the teardown method name, probably naming it `Teardown` would make the helper class even more generic.

File: test/extended/machines/scale.go
Comment (andfasano) [reply test/extended/machines/scale.go:216]: See previous comment on the usability of the helper class

**Review by elfosardo — APPROVED** [review]

File: test/extended/baremetal/helper.go
Comment (bfournie) [inline test/extended/baremetal/helper.go:128]: Wondering if its intentional to use context.Background() here and context.TODO() in other places (e.g. line 149)?

File: test/extended/machines/scale.go
Comment (bfournie) [inline test/extended/machines/scale.go:211]: This will only deploy one extra worker, is a mechanism needed to deploy more than one?

File: test/extended/machines/scale.go
Comment (bfournie) [inline test/extended/machines/scale.go:218]: When scaling down the number of replicas, would it be possible to define that a particular extra worker should be scaled down using the equivalent of "oc annotate machine <extra-worker> machine.openshift.io/cluster-api-delete-machine=yes"? Perhaps the test is already doing that and I missed it.

Comment (andfasano) [reply test/extended/machines/scale.go:218]: Yes @bfournie the test is exactly using that approaching, by marking the newly created machine with the `cluster-api-delete-machine` annotation (see [here](https://github.com/openshift/origin/blob/838021cc5dfa1a75b6efacfc77ecae65850f564e/test/extended/machines/scale.go#L260-L262))

**Comment by ardaguclu** [comment]
e2e-metal-ipi-serial is perm failing. Fix should be https://github.com/openshift/release/pull/25304

**Comment by ardaguclu** [comment]
@andfasano after this https://github.com/openshift/origin/pull/26660 is merged, there is merge conflict.

**Comment by andfasano** [comment]
> @andfasano after this #26660 is merged, there is merge conflict.

Thanks, done. @ardaguclu @bfournie could you please have another look?

**Comment by bparees** [comment]
> The existing scale test has been also retouched to adopt the new add worker mechanism in case of baremetal ipi platform

how was this test passing on baremetal ipi before, if the additional worker is needed?

**Comment by andfasano** [comment]
> > The existing scale test has been also retouched to adopt the new add worker mechanism in case of baremetal ipi platform
> 
> how was this test passing on baremetal ipi before, if the additional worker is needed?

Previously the extra worker was directly created by the dev-scripts setup via the [APPLY_EXTRA_WORKERS](https://github.com/openshift/release/blob/7a8dd52be2f1d9418102be47762d19572ffe9cb7/ci-operator/step-registry/baremetalds/e2e/serial/ipv4/baremetalds-e2e-serial-ipv4-workflow.yaml#L15) config variable (so it was available just at the end of the setup phase, before the e2e test step).
We plan to remove the `APPLY_EXTRA_WORKERS` as soon as this PR lands (current code also takes care of the migration)

**Pull request merged**
